Content.Shared/Roles/SharedRoleTimerSystem.cs

```diff
@@ -0,0 +1,111 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Robust.Shared.Configuration;
+using Robust.Shared.Network;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Roles;
+
+public abstract class SharedRoleTimerSystem : EntitySystem
+{
+    [Dependency] protected readonly IConfigurationManager ConfigManager = default!;
+    [Dependency] protected readonly IPrototypeManager ProtoManager = default!;
+
+    protected ISawmill Sawmill = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        Sawmill = Logger.GetSawmill("roletimers");
+    }
+
+    public bool TryRequirementMet(NetUserId id, JobPrototype job,
+        ref TimeSpan? overallTime,
+        ref Dictionary<string, TimeSpan>? roleTimes,
+        [NotNullWhen(false)] out string? reason)
+    {
+        reason = null;
+        if (job.Requirements == null) return true;
+
+        foreach (var requirement in job.Requirements)
+        {
+            if (!TryRequirementMet(id, requirement, ref overallTime, ref roleTimes, out reason)) return false;
+        }
+
+        return true;
+    }
+
+    /// <summary>
+    /// Returns a string with the reason why a particular requirement may not be met.
+    /// </summary>
+    public bool TryRequirementMet(NetUserId id,
+        JobRequirement requirement,
+        ref TimeSpan? overallTime,
+        ref Dictionary<string, TimeSpan>? roleTimes,
+        [NotNullWhen(false)] out string? reason)
+    {
+        reason = null;
+
+        switch (requirement)
+        {
+            case DepartmentTimeRequirement deptRequirement:
+                roleTimes ??= GetRolePlaytimes(id);
+                var playtime = TimeSpan.Zero;
+
+                // Check all jobs' departments
+                var jobs = ProtoManager.Index<DepartmentPrototype>(deptRequirement.Department).Roles.ToHashSet();
+                string proto;
+
+                // Check all jobs' playtime
+                foreach (var other in jobs)
+                {
+                    // The schema is stored on the Job role but we want to explode if the timer isn't found anyway.
+                    proto = "Job" + other;
+                    ProtoManager.Index<RoleTimerPrototype>(proto);
```

Comment:
This is still pretty icky. Because there's no yaml linting, there should at least be a test that ensures that every job prototype has a corresponding "Job<ID>" role prototype.